# PROYECTO FINAL

## LIBRERIAS

### IMPORTACION DE LIBRERIAS

In [2]:
import pandas as pd
import numpy as np
import os

## FUNCIONES

### CAMBIO CAMPO A TIPO DATATIME

In [4]:
def CambioTipoFecha(df,columnas): # Recibimos dos argumentos, los cuales son: el dataframe a modificar y las coolumnas a modificar
    df[columnas] = df[columnas].apply(pd.to_datetime) # aplicamos a las columnas obtenidas el cambio de tipo de dato a datatime
    df[columnas] = df[columnas].apply(lambda x: x.dt.date)
    return df # Retornamos el dataframe con el cambio de tipo de dato en las clumnas 

### CAMBIO CAMPO A MINUSCULA

In [6]:
def CambioMinuscula(df,columnas): # Recibimos dos argumentos, los cuales son: el dataframe a modificar y las coolumnas a modificar
    for columna in columnas: # trabajamos con un for para seleccionar columna por columna deseadas, ya que str.lower trabaja por serie y no por dataframe
        df[columna] = df[columna].str.lower() # aplicamos str.lower, el cual hace que todos los valores aparezcan en minuscula
    return df # retornamos el dataframe modificado

### CAMBIO CAMPO A MAYUSCULA

In [8]:
def CambioMayuscula(df,columnas): # Recibimos dos argumentos, los cuales son: el dataframe a modificar y las coolumnas a modificar
    for columna in columnas: # trabajamos con un for para seleccionar columna por columna deseadas, ya que str.lower trabaja por serie y no por dataframe
        df[columna] = df[columna].str.upper() # aplicamos str.upper, el cual hace que todos los valores aparezcan en mayuscula 
    return df # retornamos el dataframe modificado

### CAMBIO CAMPO A TIPO STR

In [10]:
def CambioTipoStr(df,columnas): # Recibimos dos argumentos, los cuales son: el dataframe a modificar y las coolumnas a modificar
    df[columnas] = df[columnas].astype("string") # aplicamos a las columnas obtenidas el cambio de tipo de dato a str
    return df # Retornamos el dataframe con el cambio de tipo de dato en las clumnas

## EXTRACCIÓN

### CARGA

In [12]:
# En todas estas lineas, lo que estamos haciendo es leer harchivos con formato CSV, los cuales se cuentran en una carpeta de un repositorio
# de github, y guardarlos cono dataframe mediante la libre ria pandas 

df_customers = pd.read_csv("https://raw.githubusercontent.com/Marrelli-Leandro/recursos-analisis-datos/main/14_Datasets_Originales/ecommerce_customers_dataset.csv")
df_orders = pd.read_csv("https://raw.githubusercontent.com/Marrelli-Leandro/recursos-analisis-datos/main/14_Datasets_Originales/ecommerce_orders_dataset.csv")
df_order_items = pd.read_csv("https://raw.githubusercontent.com/Marrelli-Leandro/recursos-analisis-datos/main/14_Datasets_Originales/ecommerce_order_items_dataset.csv")
df_products = pd.read_csv("https://raw.githubusercontent.com/Marrelli-Leandro/recursos-analisis-datos/main/14_Datasets_Originales/ecommerce_products_dataset.csv")
df_order_payments = pd.read_csv("https://raw.githubusercontent.com/Marrelli-Leandro/recursos-analisis-datos/main/14_Datasets_Originales/ecommerce_order_payments_dataset.csv")

### VISUALIZACIÓN

REALIZAREMOS UNA VISUALIZACION DE LOS 5 DATAFRAME, CON EL FIN DE PODER OBSERVAR

QUE TIPO DE DATOS ENCONTRAMOS EN CADA UNO DE ELL0S

#### DATAFRAME DF_CUSTOMERS

In [170]:
df_customers.head() # Visalizamos los primeros 5 registros del dataframe df_customers, con el fin de poder ver que campos contiene y tipos de datos

customer_id                customer_unique_id  \
0  06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1  18955e83d337fd6b2def6b18a428ac77  290c77bc529b7ac935b93aa66c333dc3   
2  4e7b3e00288586ebd08712fdd0374a03  060e732b5b29e8181a18229c7b0b2b5e   
3  b2b6027bc5c5109e529d4dc6358b12c3  259dac757896d24d7702b9acbbff3f3c   
4  4f2d8ab171c80ec8364f7c12e35b23ad  345ecd01c38d18a9036ed96c73b8d066   

   customer_zip_code_prefix          customer_city customer_state  
0                     14409                 franca             SP  
1                      9790  sao bernardo do campo             SP  
2                      1151              sao paulo             SP  
3                      8775        mogi das cruzes             SP  
4                     13056               campinas             SP

In [174]:
df_customers.shape # Con shape lo que hacemos, es visualizar la estructura del dataframe, con el fin de ver cuantos registros y campos hay

(99441, 5)

In [176]:
df_customers.dtypes # Con dtypes, visualizamos el tipo de dato que encontramos en cada campo

customer_id                 object
customer_unique_id          object
customer_zip_code_prefix     int64
customer_city               object
customer_state              object
dtype: object

In [178]:
df_customers.isnull().sum() # con isnull.sum, realizamos un conteo de cantidad de valores nulos en cada campo

customer_id                 0
customer_unique_id          0
customer_zip_code_prefix    0
customer_city               0
customer_state              0
dtype: int64

In [180]:
df_customers.duplicated().sum() # Con duplicated.sum, realizamos un conteo de valores duplicados

0

PODEMOS OBSERVAR EN EL DATAFRAME DF_CUSTOMERS, QUE PRESETA 5 CAMPOS Y 99441 REGISTROS

TENEMOS CAMPOS DE TIPO OBJETO Y DE TIPO ENTERO, LOS CUALES MAS ADELANTE SERA ESTUDIADO

SI ES NECESARIO ESTABLECER UN NUEVO TIPO DE VARIABLE

POR ULTIMO PODES INFORMAR QUE NO PRESENTA DATOS NULOS, POR LO CUAL NO SERA NECESARIO TRATAR DICHO TEMA Y TAMPOCO CONTIENE REGISTROS DUPLICADOS

#### DATAFRAME DF_ORDERS

In [183]:
df_orders.head() # visualizamos los primero 5 registros del dataframe df_orders, con el fin de ver que campos y tipo de datos hay

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3  949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4  ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   

  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
1    delivered      2018-07-24 20:41:37  2018-07-26 03:24:27   
2    delivered      2018-08-08 08:38:49  2018-08-08 08:55:23   
3    delivered      2017-11-18 19:28:06  2017-11-18 19:45:59   
4    delivered      2018-02-13 21:18:39  2018-02-13 22:20:29   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2018-07-26 14:31:00           2018-08-07 15:27:45   
2          2018-08-08 13:50:00           2018-08-17 18:06:29   
3          2017-11-22 13:39:59           2017-12-02 00:28:42   
4          2018-02-14 19:46:34           2018-02-16 18:17:02   

  order_estimated_delivery_date  
0           2017-10-18 00:00:00  
1           2018-08-13 00:00:00  
2           2018-09-04 00:00:00  
3           2017-12-15 00:00:00  
4           2018-02-26 00:00:00

In [185]:
df_orders.shape # con shape vemos la estructura del dataframe, pasa saber cuantos registros y campos hay

(99441, 8)

In [187]:
df_orders.dtypes # con dtype vemos que tipos de datos tenemos en los campos del dataframe

order_id                         object
customer_id                      object
order_status                     object
order_purchase_timestamp         object
order_approved_at                object
order_delivered_carrier_date     object
order_delivered_customer_date    object
order_estimated_delivery_date    object
dtype: object

In [189]:
df_orders.isnull().sum() # con isnull.sum podemos ver la cantidad de valores nulos por cada campo

order_id                            0
customer_id                         0
order_status                        0
order_purchase_timestamp            0
order_approved_at                 160
order_delivered_carrier_date     1783
order_delivered_customer_date    2965
order_estimated_delivery_date       0
dtype: int64

In [191]:
df_orders.duplicated().sum() # Con duplicated.sum podemos visualizar cuantos calores duplicados hay

0

PODEMOS OBSERVAR EN EL DATAFRAME DF_ORDERS, QUE PRESETA 8 CAMPOS Y 99441 REGISTROS

TENEMOS TODOS LOS CAMPOS DE TIPO OBJETO, PERO SE OBSERVAN REGISTOS DE FECHAS, POR LO QUE SERA NECESARIO

PASARLOS A TIPO DATE. TAMBIEN SERA ESTUDIADO SI ES NECESARIO ESTABLECER UN NUEVO TIPO DE VARIABLE PARA LOS CAMPOS RESTANTES

POR ULTIMO PODEMOS INFORMAR QUE PRESENTA REGISTROS NULOS, EN TRES CAMPOS, LOS CUAL DEBERÁ SER TRATADO MAS ADELANTE, EVALUANDO CUAL ES LA ACCIÓN

MAS CONVENIENTE PARA ESTE CASO Y TAMBIÉN PODEMOS DECIR QUE NO PRESENTA DUPLICADOS

#### DATAFRAME DF_ORDER_ITEMS

In [194]:
df_order_items.head() # Visualizamos los primeros 5 registros del dataframe df_order_items, con el fin de poder ver que camposy datos tenemos

order_id  order_item_id  \
0  00010242fe8c5a6d1ba2dd792cb16214              1   
1  00018f77f2f0320c557190d7a144bdd3              1   
2  000229ec398224ef6ca0657da4fc703e              1   
3  00024acbcdf0a6daa1e931b038114c75              1   
4  00042b26cf59d7ce69dfabb4e55b4fd9              1   

                         product_id                         seller_id  \
0  4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   
1  e5f2d52b802189ee658865ca93d83a8f  dd7ddc04e1b6c2c614352b383efe2d36   
2  c777355d18b72b67abbeef9df44fd0fd  5b51032eddd242adc84c38acab88f23d   
3  7634da152a4610f1595efa32f14722fc  9d7a1d34a5052409006425275ba1c2b4   
4  ac6c3623068f30de03045865e4e10089  df560393f3a51e74553ab94004ba5c87   

   shipping_limit_date   price  freight_value  
0  2017-09-19 09:45:35   58.90          13.29  
1  2017-05-03 11:05:13  239.90          19.93  
2  2018-01-18 14:48:30  199.00          17.87  
3  2018-08-15 10:10:18   12.99          12.79  
4  2017-02-13 13:57:51  199.90          18.14

In [196]:
df_order_items.shape # Con shape estamos viendo la estructura del dataframe para saber cuando registros y campos hay

(112650, 7)

In [198]:
df_order_items.dtypes # Con dtype estamos viendo que tipos de datos tenemos en el dataframe

order_id                object
order_item_id            int64
product_id              object
seller_id               object
shipping_limit_date     object
price                  float64
freight_value          float64
dtype: object

In [200]:
df_order_items.isnull().sum() # en esta linea estamos contando la cantidad de valores nuloes q tenemos en cada campo

order_id               0
order_item_id          0
product_id             0
seller_id              0
shipping_limit_date    0
price                  0
freight_value          0
dtype: int64

In [202]:
df_order_items.duplicated().sum() # En esta lonea estamos visualizando cuando duplicados existen

0

PODEMOS OBSERVAR EN EL DATAFRAME DF_ORDER_ITEMS, QUE PRESETA 7 CAMPOS Y 112650 REGISTROS

TENEMOS CAMPOS DE TIPO OBJETO, DE TIPO ENTERO Y DE TIPO FLOAT, LOS CUALES MAS ADELANTE SERA ESTUDIADO

SI ES NECESARIO ESTABLECER UN NUEVO TIPO DE VARIABLE. YA SE PUEDE DECIR QUE EL CAMPO QUE CONTIENE FECHAS, QUE ES DE TIPO OBJETO

TENDRA QUE SER SETEADO A TIPO DATE

POR ULTIMO PODEMOS INFORMAR QUE NO PRESENTA REGISTROS NULOS Y TAMPOCO TIENE DUPLICADOS

#### DATAFRAME DF_PRODUCTS

In [205]:
df_products.head() # Con head visualizamos los primeros 5 registros para poder ver que campos y q variables nos encontrmos

product_id  product_category_name  \
0  1e9e8ef04dbcff4541ed26657ea517e5             perfumaria   
1  3aa071139cb16b67ca9e5dea641aaa2f                  artes   
2  96bd76ec8810374ed1b65e291975717f          esporte_lazer   
3  cef67bcfe19066a932b7673e239eb23d                  bebes   
4  9dc1a7de274444849c219cff195d0b71  utilidades_domesticas   

   product_name_lenght  product_description_lenght  product_photos_qty  \
0                 40.0                       287.0                 1.0   
1                 44.0                       276.0                 1.0   
2                 46.0                       250.0                 1.0   
3                 27.0                       261.0                 1.0   
4                 37.0                       402.0                 4.0   

   product_weight_g  product_length_cm  product_height_cm  product_width_cm  
0             225.0               16.0               10.0              14.0  
1            1000.0               30.0               18.0              20.0  
2             154.0               18.0                9.0              15.0  
3             371.0               26.0                4.0              26.0  
4             625.0               20.0               17.0              13.0

In [207]:
df_products.shape # en esta linea buscamos visualizar cual es la estructura del dataframe, para saber cuantos campos y regustros existen

(32951, 9)

In [209]:
df_products.dtypes # con dtypes buscamos saber que tipo de variables existen en cada campo

product_id                     object
product_category_name          object
product_name_lenght           float64
product_description_lenght    float64
product_photos_qty            float64
product_weight_g              float64
product_length_cm             float64
product_height_cm             float64
product_width_cm              float64
dtype: object

In [211]:
df_products.isnull().sum() # visualizamos la cantidad de valores nulos en cada campo

product_id                      0
product_category_name         610
product_name_lenght           610
product_description_lenght    610
product_photos_qty            610
product_weight_g                2
product_length_cm               2
product_height_cm               2
product_width_cm                2
dtype: int64

In [213]:
df_products.duplicated().sum() # de esta manera contamos la cantidad de duplicados

0

PODEMOS OBSERVAR EN EL DATAFRAME DF_PRODUCTS, QUE PRESENTA 9 CAMPOS Y 32951 REGISTROS

TENEMOS CAMPOS DE TIPO OBJETO Y DE TIPO FLOAT, LOS CUALES MAS ADELANTE SERAN ESTUDIADO

SI ES NECESARIO ESTABLECER UN NUEVO TIPO DE VARIABLE

POR ULTIMO PODEMOS VER QUE EL DATAFRAME PRESENTA REGISTROS NULOS EN TODOS SUS CAMPOS, MENOS EN UNO, LO CUAL SERA TRATADO MAS ADELANTE,

APLICANDO EL METODO MAS CONVENIENTE PARA ESTE CASO Y TAMBIEN PODEMOS INFORMA QUE NO PRESENTA DUPLICADOS

#### DATAFRAME DF_ORDER_PAYMENTS

In [216]:
df_order_payments.head() # En esta linea buscamos visualizar los primeros 5 registros, con el fin de ver los campos y q variables existen

order_id  payment_sequential payment_type  \
0  b81ef226f3fe1789b1e8b2acac839d17                   1  credit_card   
1  a9810da82917af2d9aefd1278f1dcfa0                   1  credit_card   
2  25e8ea4e93396b6fa0d3dd708e76c1bd                   1  credit_card   
3  ba78997921bbcdc1373bb41e913ab953                   1  credit_card   
4  42fdf880ba16b47b59251dd489d4441a                   1  credit_card   

   payment_installments  payment_value  
0                     8          99.33  
1                     1          24.39  
2                     1          65.71  
3                     8         107.78  
4                     2         128.45

In [218]:
df_order_payments.shape # En esta linea se consulta cual es la estructura del dataframe, con el fin de ver la cantidad de campos y registros

(103886, 5)

In [220]:
df_order_payments.dtypes # con dtypes consultamos el tipo de variable que existe en cada campo

order_id                 object
payment_sequential        int64
payment_type             object
payment_installments      int64
payment_value           float64
dtype: object

In [222]:
df_order_payments.isnull().sum() # aca estamos contando la cantidad de valores nulos que existen en cada campo

order_id                0
payment_sequential      0
payment_type            0
payment_installments    0
payment_value           0
dtype: int64

In [224]:
df_order_payments.duplicated().sum() # con esta linea contamos la cantidad de duplicados

0

PODEMOS OBSERVAR EN EL DATAFRAME DF_ORDER_PAYMENTS, QUE PRENSETA 5 CAMPOS Y 103886 REGISTROS

TENEMOS CAMPOS DE TIPO OBJETO, DE TIPO ENTERO Y TIPO FLOAT, LOS CUALES MAS ADELANTE SERA ESTUDIADO

SI ES NECESARIO ESTABLECER UN NUEVO TIPO DE VARIABLE

POR ULTIMO PODEMOS INFORMAR QUE NO SE PRESENTAN REGISTROS NULOS Y TAMPOCO DUPLICADOS.

## TRANSFORMACIÓN

###  DF_ORDERS

#### ANALIZAMOS VALORES NULOS

PARA EL TRATAMIENTO DE LOS NULOS EN ESTE DATAFRAME, TENDREMOS EN CONCIDERACION QUE MANTENDREMOS AQUELAS COLUMNAS QUE NO SUPEREN

UN 5% DE VALORES NULOS, EN EL TOTAL DE VALORES. UNA VEZ QUE TENGAMOS EL DATAFRAME NUEVO CON LAS COLUMNAS QUE COMPLAN CON DICHA

CONDICION, SE PROCEDERÁ A REALIZAR LA ELIMINACION DE LOS REGISTROS QUE CONTENGAN VALORES NULOS.

In [14]:
# En la linea siguiente lo que estamos haciendo es obtener de manera porcentual, respecto a la antidad total de registros, cual es el
# porcentaje de valore nulos en cada campo.
porcentaje_nulos = (df_orders.isnull().sum())*100/(len(df_orders))
porcentaje_nulos

# Podemos ver cuales son los porcentajes de nulosque existe en cada campo. En la linea siguiente vamos a filtrar por porcentaje, es decir
# los camposque tengan un porcentaje de nulos menor a 5% seran campos considerados, si supera al 5%, esos campos no seran tenidos en cuenta.
# Los campos se encuentran en el index, es por eso que estamos guardando en columnas, las columnas que complen con la condicion explicada.
# Elcampo que complua con la condicion del 5% y contenga valores nulos, esos registros seran eliminados
columnas = porcentaje_nulos[porcentaje_nulos  < 5].index
df_orders = df_orders[columnas] # en esta linea estamos modificando el dataframe para que contenga las clumnas deseadas
df_orders = df_orders.dropna() # en esta linea estamos eliminado los registros que contienen valores nulos.

#### CAMBIAMOS CAMPOS A TIPO DATATIME

In [15]:
# En esta celda vamos a convertir el tipo de datos de algunos campos a tipo datatime, para eso en la siguiente linea lo que hcemos es
# crear una lista con los campos deseados.
columnas = ["order_purchase_timestamp","order_approved_at","order_delivered_carrier_date","order_delivered_customer_date","order_estimated_delivery_date"]
# en la siguiente linea llamamos a la funcon que se encarga de de realizar el cambio de tipo de variable, en la que le enviamos el dataframe que se quiere
# modificar y tambien la lista con las columnas que deseamos modificar
df_orders_mod = CambioTipoFecha(df_orders,columnas)

#### CAMBIAMOS CAMPOS A TIPO STR

In [20]:
# En esta celda lo que hacemos es llamar a la funcion que se encarga de realizar el cambio de tipo de dato a tipo str. Para eso en la siguiente linea
# lo que hacemos es crear una lista con las columnas que deseamos realizar dicho cambio.
columnas = ["order_id","customer_id","order_status"]
# Una vez que se tenga la lista con los campos deseado, en la siguiente linea llamamos a la funcion y guardamos en df_orders_mod, el dataframe con los
# cambios realizados por la funcion.
df_orders_mod = CambioTipoStr(df_orders_mod,columnas)

#### CAMBIO CAMPO A MINUSCULA

In [22]:
# En esta celda lo que se hace es llamar a la funcion que se encarga de hacer que todos los calores de un campo que contriene str, aparezcan
# con minuscula. Para eso en la siguiente linea lo que se hace es crear una lista con los campos que deseamos que aparezcan en minuscula.
columnas = ["order_status"]
# Una vez que ya tenemos la lista con los campos, lo que hacemos es llamar a la funcion que se encarga de eso, donde se le envia el dataframe
# y la lista con los campos que deseamos modificar.
df_orders_mod = CambioMinuscula(df_orders_mod,columnas)

In [24]:
df_orders_mod.dtypes
# con esta linea podemos visualizar cuales son los tipos de datos de los campos del dataframe que ya fue modificado,
# donde podemos ver que ahora tenemos los campos con los tipos de variables deseados

order_id                         string[python]
customer_id                      string[python]
order_status                     string[python]
order_purchase_timestamp                 object
order_approved_at                        object
order_delivered_carrier_date             object
order_delivered_customer_date            object
order_estimated_delivery_date            object
dtype: object

In [26]:
df_orders_mod.isnull().sum()
# Con esta linea buscamos verificar que no existan mas valores nulos, y realmente es posible verificar que en eldataframe
# ya no existen valores nulos, que es lo que se buscaba

order_id                         0
customer_id                      0
order_status                     0
order_purchase_timestamp         0
order_approved_at                0
order_delivered_carrier_date     0
order_delivered_customer_date    0
order_estimated_delivery_date    0
dtype: int64

In [28]:
df_orders_mod.shape
# De esta manera podemos ver que el dataframe sigue conservando la cantidad de campos, pero ya no contiene la misma cantidad de registros,
# ya que se han eliminados varios registros.

(96461, 8)

In [32]:
df_orders_mod.head()
# en esta linea volvemos a visualizar los primeros 5 registros del dataframe modificado, con el fin de poder visualizar
# lo cambios realizados.

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3  949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4  ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   

  order_status order_purchase_timestamp order_approved_at  \
0    delivered               2017-10-02        2017-10-02   
1    delivered               2018-07-24        2018-07-26   
2    delivered               2018-08-08        2018-08-08   
3    delivered               2017-11-18        2017-11-18   
4    delivered               2018-02-13        2018-02-13   

  order_delivered_carrier_date order_delivered_customer_date  \
0                   2017-10-04                    2017-10-10   
1                   2018-07-26                    2018-08-07   
2                   2018-08-08                    2018-08-17   
3                   2017-11-22                    2017-12-02   
4                   2018-02-14                    2018-02-16   

  order_estimated_delivery_date  
0                    2017-10-18  
1                    2018-08-13  
2                    2018-09-04  
3                    2017-12-15  
4                    2018-02-26

### DF_CUSTOMERS

#### CAMBIAMOS CAMPOS A TIPO STR

In [49]:
# En esta celda lo que buscamos hacer es cambiar el tipo de datos de algunas columnas en tipo str. En la siguiente linea
# creamos una lista con las columnas a las cuales neceitamos realizar dicho cambio
columnas = ["customer_id","customer_unique_id","customer_city","customer_state"]
# En la linea siguiente llamamos a la funcion encargada de dicha tarea, en donde enviamos el dataframe que deceamos modifical
# y tambien las columnas de dicho dataframe qye queremos cambiar
df_customers_mod = CambioTipoStr(df_customers,columnas)

#### CAMBIO CAMPO A MINUSCULA

In [51]:
# En la sguiente celda buscamos modificar columnas que son str y que aparezcan todos los registros escritos en minuscula
# en la linea siguiente creamos una lista con las columnas que necesitamos modificar
columnas = ["customer_city"]
# ahora llamamos a la funcion encargada de eso yle enviamos el dataframe que necesitamos modificar, con las
# columnas necesarias
df_customers_mod = CambioMinuscula(df_customers_mod,columnas)

#### CAMBIO CAMPO A MAYUSCULA

In [53]:
# En la sguiente celda buscamos modificar columnas que son str y que aparezcan todos los registros escritos en mayuscula
# en la linea siguiente creamos una lista con las columnas que necesitamos modificar
columnas = ["customer_state"]
# ahora llamamos a la funcion encargada de eso yle enviamos el dataframe que necesitamos modificar, con las
# columnas necesarias
df_customers_mod = CambioMayuscula(df_customers_mod,columnas)

In [55]:
# con dtypes vamos a controlar que se hayan realizado los cambios de tipo de datos. 
df_customers_mod.dtypes

customer_id                 string[python]
customer_unique_id          string[python]
customer_zip_code_prefix             int64
customer_city               string[python]
customer_state              string[python]
dtype: object

In [57]:
# visualmente controlamos los primeros 5 registros del dataframe, con el fin de podecon controlar si se hicieron
# los cambios anteriores
df_customers_mod.head()

customer_id                customer_unique_id  \
0  06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1  18955e83d337fd6b2def6b18a428ac77  290c77bc529b7ac935b93aa66c333dc3   
2  4e7b3e00288586ebd08712fdd0374a03  060e732b5b29e8181a18229c7b0b2b5e   
3  b2b6027bc5c5109e529d4dc6358b12c3  259dac757896d24d7702b9acbbff3f3c   
4  4f2d8ab171c80ec8364f7c12e35b23ad  345ecd01c38d18a9036ed96c73b8d066   

   customer_zip_code_prefix          customer_city customer_state  
0                     14409                 franca             SP  
1                      9790  sao bernardo do campo             SP  
2                      1151              sao paulo             SP  
3                      8775        mogi das cruzes             SP  
4                     13056               campinas             SP

### DF_ORDER_ITEMS

#### CAMBIAMOS CAMPOS A TIPO DATATIME

In [59]:
# en esta celda llamamos a la funcion encargada de cambiar el tipo de dato de las columnas a tipo datatime
# en la siguiente linea creamos una lista con las columas que necesitamos realizar dicho cambio
columnas = ["shipping_limit_date"]
# ahora llamamos a la funcion y le enviamos el dataframe y las columnas que deseanis cambiar
df_order_items_mod = CambioTipoFecha(df_order_items,columnas)

#### CAMBIAMOS DATOS A TIPO STR

In [61]:
# en esta celda llamanos a la funcion encargada de cambiar el tipo de datos de las columnas a str
# en la siguiente line creamos una lista con las columnas que necesitamos cambiar
columnas = ["order_id","product_id","seller_id"]
# ahora llamamos a la funcion y le enviamos el dataframe necesario con las columnas que deseamos cambiar
df_order_items_mod = CambioTipoStr(df_order_items_mod,columnas)

In [63]:
# controlamos que los cambios ahora tengas los tipos de datos que necesitamos
df_order_items_mod.dtypes

order_id               string[python]
order_item_id                   int64
product_id             string[python]
seller_id              string[python]
shipping_limit_date            object
price                         float64
freight_value                 float64
dtype: object

### DF_PRODUCTS

#### ANALIZAMOS VALORES NULOS

ESTE DATAFRAME POSEE 4 COLUMAS CON VALORES NULOS, LAS CUALES SON LAS SIGUINTES:

product_category_name         610

product_name_lenght           610

product_description_lenght    610

product_photos_qty            610

product_weight_g                2

product_length_cm               2

product_height_cm               2

product_width_cm                2

LA REGLA QUE SERA UTILIZADA EN EL MANEJO DE LOS NULOS ES LA SIGUIENTE:

PARA EL CAMPO "product_category_name" REEMPLAZAREMOS LOS CAMPOS NULS POR "Desconocida"

PARA EL RESTO DE LOS CAMPOS, REEMPLAZAREMOS LOS NULOS POR EL VALOR "-1", EL CUAL DICHO VALOR, INDICA QUE ES DESCONOCIDO EL VALOR REAL.

SE TOMO ESTA DECICION, YA QUE LOS CAMPOS CON LOS QUE ESTAMOS TRABAJANDO, NO SON DE VALORES DE PRECIO, NI CANTIDAD DE STOCK, NI OTROS VALORES 

DE ESE TIPO. LOS VALORES CON LOS QUE TRABAJAMOS EN ESTE DATAFRAME, SON TAMAÑO DE PRODUCTO, LARGO DEL NOMBRE, CANTIDAD DE IMAGENES EN EL POSTEO.

SE TENDRA EN CUENTA QUE AL MOMENTO DE QUERER REALIZAR ALGUN CALCULO, DE SER NECESARIO EXCLUIREMOS A LOS REGISTROS SON -1, YA QUE SON DESCONOCIDOS

In [65]:
# en la siguiente line vamos a completar los vlores nulos con la palabra "Desconocido"
df_products["product_category_name"] = df_products["product_category_name"].fillna("Desconocida")
# A continuacion creamos una lista con las columnas que vamos a querer modificar, luego a todas esas columnas la completamos
#en sus valores nulos con el valor -1
columnas = ["product_name_lenght","product_description_lenght","product_photos_qty","product_weight_g","product_length_cm","product_height_cm","product_width_cm"]
df_products[columnas] = df_products[columnas].fillna(-1)

#### CAMBIAMOS CAMPOS A TIPO STR

In [67]:
# en esta celda lo que hacemos es llamar a la funcion encargada de cambiar el tipo de dato de los campos a tipo str.
# en la siguiente linea creamos una lista con los campos que desemaos cambiar el tipo de dato a tipo str
columnas = ["product_id","product_category_name"]
# luego llamamos a la funcion que se encarda de dicha tarea y le eviamos el dataframe y las columnas que deseamos cambiar
df_products_mod = CambioTipoStr(df_products,columnas)

#### CAMBIAMOS CAMPO A MINUSCULA

In [69]:
# en esta celda estamos buscando que en los campos deseados del dataframe, nos aseguremos que aparezcan en minuscula.
# en la proxima linea creamos una lista con los campos deaseados a cambiar a minuscula
columnas = ["product_category_name"]
# ahora llamanos a la funcion encargada de dicha tarea y le enviamos el dataframe y las columnas que deseamos cambiar
df_products_mod = CambioMinuscula(df_products_mod,columnas)

In [71]:
# visualizamos que los campos ahora tengan los tipos de datos que necesitamos
df_products_mod.dtypes

product_id                    string[python]
product_category_name         string[python]
product_name_lenght                  float64
product_description_lenght           float64
product_photos_qty                   float64
product_weight_g                     float64
product_length_cm                    float64
product_height_cm                    float64
product_width_cm                     float64
dtype: object

In [73]:
#controlamos que la sumatoria de valores nulos en cada campo sea cero
df_products_mod.isnull().sum()

product_id                    0
product_category_name         0
product_name_lenght           0
product_description_lenght    0
product_photos_qty            0
product_weight_g              0
product_length_cm             0
product_height_cm             0
product_width_cm              0
dtype: int64

In [75]:
# vemos la estructura del dataframe y se observa que la cantidad de columnas sigue siendo la mismoa, al igual que los registros,
# ya que no se eliminaron registros, si no que se completaron con otros valres
df_products_mod.shape

(32951, 9)

In [77]:
# controlamos visualmente que los datos aparezcan como los hemos trabajado
df_products_mod.head()

product_id  product_category_name  \
0  1e9e8ef04dbcff4541ed26657ea517e5             perfumaria   
1  3aa071139cb16b67ca9e5dea641aaa2f                  artes   
2  96bd76ec8810374ed1b65e291975717f          esporte_lazer   
3  cef67bcfe19066a932b7673e239eb23d                  bebes   
4  9dc1a7de274444849c219cff195d0b71  utilidades_domesticas   

   product_name_lenght  product_description_lenght  product_photos_qty  \
0                 40.0                       287.0                 1.0   
1                 44.0                       276.0                 1.0   
2                 46.0                       250.0                 1.0   
3                 27.0                       261.0                 1.0   
4                 37.0                       402.0                 4.0   

   product_weight_g  product_length_cm  product_height_cm  product_width_cm  
0             225.0               16.0               10.0              14.0  
1            1000.0               30.0               18.0              20.0  
2             154.0               18.0                9.0              15.0  
3             371.0               26.0                4.0              26.0  
4             625.0               20.0               17.0              13.0

### DF_ORDER_PAYMENTS

#### CAMBIAMOS CAMPOS A TIPO STR

In [79]:
# en esta caleda llamamos a la funcion encargada de transformar el tipo de dato de los campos de un dataframe a tipo str
# en la siguiente linea creamos una lista con los campos que neceitamos mnodificar
columnas = ["order_id","payment_type"]
# ahora llamamos a la funcion, enviandole el dataframe y los campos necesarios
df_order_payments_mod = CambioTipoStr(df_order_payments,columnas)

#### CAMBIAMOS CAMPOS A MINUSCULA

In [81]:
# en esta celda llamamos a la funcion encargada de hacer que losvalores de un campo aparezcan en minuscula
# en la linea que sigue creamos una lista con los campos que necesitamos cambiar
columnas = ["payment_type"]
# en la linea que sigue llamamos a la funcion y le enviamos el dataframe y los campos deseados
df_order_payments_mod = CambioMinuscula(df_order_payments_mod,columnas)

In [83]:
# ahora visualizamos que los tipos de datos de los campos esten como necesitamos.
df_order_payments_mod.dtypes

order_id                string[python]
payment_sequential               int64
payment_type            string[python]
payment_installments             int64
payment_value                  float64
dtype: object

### CREACION DE DATAFRAME FINALES

#### SELECCION DE COLUMNAS

In [224]:
# En esta celda lo que realizamos es dejar a cada uno de los dataframe, las columnas que vamos a utilizar

df_orders_mod = df_orders_mod[["order_id","customer_id","order_purchase_timestamp","order_delivered_customer_date","order_delivered_carrier_date"]]
df_customers_mod = df_customers_mod[["customer_id","customer_city"]]
df_order_items_mod = df_order_items_mod[["order_id","order_item_id","product_id","price"]]
df_products_mod = df_products_mod[["product_id","product_category_name"]]
df_order_payments_mod = df_order_payments_mod[["order_id","payment_type","payment_installments"]]

#### MERGE

##### DF_ORDER Y DF_CUSTOMER

In [290]:
# en la siguiente linea se realiza un merge entre df_orders_mod ydf_customer_mod mediante la columna customer_id
# y realizando un merge inner
df_customer_order = pd.merge(df_orders_mod,df_customers_mod,how = "inner", on = "customer_id")

# Con este dataframe lo que buscamos es poder ver varios puntos

'''Por un lado se va a realizar un estudio de cantidades de compras por lugar geografico, y analizar si se ejecutaran
 trabajos para reforzar mas las zonas con mayor cantidades de compras o si se realizan trabajos de marketing para zonas
 en las que hay menos ventasy se quiere lograr una mayor cantidad'''

'''Por otro lado analizaremos cuales son las demoras de entrega desde que se realiza la compra, hasta que se entrega el
 producto al cliente. Al mismo tiempo ver cual es la demora que hay desde que se realiza la compra, hasta que se le entrega
 el producto al repartidor y analizar si el problema si la mayor demora se ecuentra en la empresa para entregar el producto
 al repartidor o si el que se demora mas es la empresa de envio en tregar el producto'''

'''Por ultimo el merge que se realiza, es para luego poder hacer otro merge con order_item y poder analizar el monto de
dinero que ingresa en las mas compradoras, ya que no es lo mismo realizar ,uchas compras, a que genre mucho ingreso,
que que pueden ser muchas compras de valores muy bajos y hay otra zona que realiza menos compras, pero de montos mucho
mayor, haciendo que se generen mayores ingresos en la empresa'''

'Por ultimo el merge que se realiza, es para luego poder hacer otro merge con order_item y poder analizar el monto de\ndinero que ingresa en las mas compradoras, ya que no es lo mismo realizar ,uchas compras, a que genre mucho ingreso,\nque que pueden ser muchas compras de valores muy bajos y hay otra zona que realiza menos compras, pero de montos mucho\nmayor, haciendo que se generen mayores ingresos en la empresa'

In [292]:
# visualizamos los primeros 5 registros para verificar el merge
df_customer_order.head()

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3  949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4  ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   

  order_purchase_timestamp order_delivered_customer_date  \
0               2017-10-02                    2017-10-10   
1               2018-07-24                    2018-08-07   
2               2018-08-08                    2018-08-17   
3               2017-11-18                    2017-12-02   
4               2018-02-13                    2018-02-16   

  order_delivered_carrier_date            customer_city  
0                   2017-10-04                sao paulo  
1                   2018-07-26                barreiras  
2                   2018-08-08               vianopolis  
3                   2017-11-22  sao goncalo do amarante  
4                   2018-02-14              santo andre

In [294]:
# vemos que despues del merge no se generen valores nulos
df_customer_order.isnull().sum()

order_id                         0
customer_id                      0
order_purchase_timestamp         0
order_delivered_customer_date    0
order_delivered_carrier_date     0
customer_city                    0
dtype: int64

In [296]:
# vemos la cantidad de compras realizadas en cada zona, visualizando solo las primeras 10
ventas_city = df_customer_order.value_counts(["customer_city"]).head(10)
ventas_city

customer_city        
sao paulo                15044
rio de janeiro            6603
belo horizonte            2697
brasilia                  2071
curitiba                  1489
campinas                  1406
porto alegre              1341
salvador                  1188
guarulhos                 1143
sao bernardo do campo      911
Name: count, dtype: int64

In [297]:
ventas = len(df_customer_order)
ventas

96461

In [300]:
#pasamos a porcentaje las cantidades de compras
ventas_city = ventas_city*100/ventas
'''Se tendra en cuanta a las zonas que tengan compras superiores al 1% del total, a las que se pondran en un plan de
trabajo para reforzar y lograr aumentar el nivel de ventas, el resto de las zonas se seguiran trabajando de la misma
manera que hasta hoy, logrando mantener el nivel de ventas, pero no se gastara ni tiempo ni dinero en lograr mejorarlas'''
# filtramos auqellas zonas que tienen compras mayores o iguales al 1%, que son las que nos importan
ventas_city[ventas_city >= 1]

customer_city 
sao paulo         15.595940
rio de janeiro     6.845254
belo horizonte     2.795949
brasilia           2.146982
curitiba           1.543629
campinas           1.457584
porto alegre       1.390199
salvador           1.231586
guarulhos          1.184935
Name: count, dtype: float64

PODEMOS OBSERVAR QUE TENEMOS 9 REGIONES QUE SUPERAN EL 1% DE VENTAS TOTALES. SIENDO A LAS QUE APLICAREMOS EL EL PLAN PARA
AUMENTAR EL NIVEL DE VENTAS.

In [303]:
# ahora vamos a filtrar las registros que solo sean de las zonas estudiadas 
zonas = ["sao paulo","rio de janeiro","belo horizonte","brasilia","curitiba","campinas","porto alegre","salvador","guarulhos"]
df_customer_order = df_customer_order[df_customer_order.customer_city.isin(zonas)]

# en la siguiente linea realizamos la cuenta para saber cuants dias se demoran en entregar el pedido despues de realizar la compra
tiempo_entrega_cliente = df_customer_order["order_delivered_customer_date"] - df_customer_order["order_purchase_timestamp"]
# filtramos aquellos registros que tienen un tiempo de demora menos igual a 7 dias
tiempo_entrega_cliente = tiempo_entrega_cliente <= pd.Timedelta(days=7)
tiempo_entrega_cliente

0        False
9        False
10        True
12        True
13       False
         ...  
96441    False
96446     True
96449     True
96450    False
96455    False
Length: 32982, dtype: bool

In [305]:
# contamos cuantas entregas cumplen don dicha condicion 
cantidad_men_7 = tiempo_entrega_cliente.value_counts()
cantidad_men_7

False    19065
True     13917
Name: count, dtype: int64

In [307]:
largo = len(tiempo_entrega_cliente)
tiempo_entrega_por = cantidad_men_7 * 100 / largo
tiempo_entrega_por

False    57.804257
True     42.195743
Name: count, dtype: float64

PODEMOS VER QUE DE LAS 9 ZONAS QUE CUMPLEN CON LA CONDICION DE TENER VENTAS POR ARRIBA DEL 1%, EL 57,80 % APROXIMADAMENTE
NO CUMPLEN CON LA ESPECTATIVA DE QUE NO TARDEN MAS DE 7 DIAS, MIENTRAS QUE EL 42 % APROXIMADAMENTE SI LO CUMPLE.
VA A SER UN TEMA MUY IMPORTANTE A TRATAR SI LO QUE SE QUIERE ES AMPLIAR LA CANTIDAD DE VENTAS EN ESTAS ZONAS

In [310]:
df_customer_order["tiempo_entrega_menor_8"] = tiempo_entrega_cliente

In [312]:
df_customer_order.head()

order_id                       customer_id  \
0   e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
9   e6ce16cb79ec1d90b1da9085a6118aeb  494dded5b201313c64ed7f100595b95c   
10  34513ce0c4fab462a55830c0989c7edb  7711cf624183d843aafe81855097bc37   
12  5ff96c15d0b717ac6ad1f3d77225a350  19402a48fe860416adf93348aba37740   
13  432aaf21d85167c2c86ec9448c4e42cc  3df704f53d3f1d4818840b34ec672a9f   

   order_purchase_timestamp order_delivered_customer_date  \
0                2017-10-02                    2017-10-10   
9                2017-05-16                    2017-05-29   
10               2017-07-13                    2017-07-19   
12               2018-07-25                    2018-07-30   
13               2018-03-01                    2018-03-12   

   order_delivered_carrier_date   customer_city  tiempo_entrega_menor_8  
0                    2017-10-04       sao paulo                   False  
9                    2017-05-18  rio de janeiro                   False  
10                   2017-07-14       sao paulo                    True  
12                   2018-07-26       sao paulo                    True  
13                   2018-03-02       sao paulo                   False

In [314]:
df_customer_order = df_customer_order[["order_id","customer_id","customer_city","tiempo_entrega_menor_8"]]

In [316]:
df_customer_order

order_id                       customer_id  \
0      e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
9      e6ce16cb79ec1d90b1da9085a6118aeb  494dded5b201313c64ed7f100595b95c   
10     34513ce0c4fab462a55830c0989c7edb  7711cf624183d843aafe81855097bc37   
12     5ff96c15d0b717ac6ad1f3d77225a350  19402a48fe860416adf93348aba37740   
13     432aaf21d85167c2c86ec9448c4e42cc  3df704f53d3f1d4818840b34ec672a9f   
...                                 ...                               ...   
96441  6f46d42016226f77893ed3978d6903d0  3803547f820b1e10b7ad9c7985f06753   
96446  7cc85185736ea7f2d8bc09af7a47cab7  9e870ea67399650e8c3545a25d13e4d0   
96449  0e4b26f1fb99fc0f0472dd5f403d36d4  2e96fa6d55b9732013caed9866e2552d   
96450  c81f74e50f0496fa39716cc77cacd460  7b270ebc87c25c8404348c10ff80a80e   
96455  880675dff2150932f1601e1c07eadeeb  47cd45a6ac7b9fb16537df2ccffeb5ac   

        customer_city  tiempo_entrega_menor_8  
0           sao paulo                   False  
9      rio de janeiro                   False  
10          sao paulo                    True  
12          sao paulo                    True  
13          sao paulo                   False  
...               ...                     ...  
96441    porto alegre                   False  
96446        brasilia                    True  
96449       sao paulo                    True  
96450  belo horizonte                   False  
96455       sao paulo                   False  

[32982 rows x 4 columns]

##### DF_CUSTOMER_ORDER Y DF_ORDER_ITEM

In [319]:
df_customer_order_item = pd.merge(df_customer_order,df_order_items_mod,how = "inner", on = "order_id")

In [321]:
df_customer_order_item.head()

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  e6ce16cb79ec1d90b1da9085a6118aeb  494dded5b201313c64ed7f100595b95c   
2  e6ce16cb79ec1d90b1da9085a6118aeb  494dded5b201313c64ed7f100595b95c   
3  34513ce0c4fab462a55830c0989c7edb  7711cf624183d843aafe81855097bc37   
4  5ff96c15d0b717ac6ad1f3d77225a350  19402a48fe860416adf93348aba37740   

    customer_city  tiempo_entrega_menor_8  order_item_id  \
0       sao paulo                   False              1   
1  rio de janeiro                   False              1   
2  rio de janeiro                   False              2   
3       sao paulo                    True              1   
4       sao paulo                    True              1   

                         product_id  price  
0  87285b34884572647811a353c7ac498a  29.99  
1  08574b074924071f4e201e151b152b4e  99.00  
2  08574b074924071f4e201e151b152b4e  99.00  
3  f7e0fa615b386bc9a8b9eb52bc1fff76  98.00  
4  10adb53d8faa890ca7c2f0cbcb68d777  19.90

In [323]:
# verificamos que el nuevo df no contenga valores nulos
df_customer_order_item.isnull().sum()

order_id                  0
customer_id               0
customer_city             0
tiempo_entrega_menor_8    0
order_item_id             0
product_id                0
price                     0
dtype: int64

In [325]:
df_customer_order_item["price"].describe()

count    37997.000000
mean       114.275758
std        172.130391
min          1.200000
25%         38.900000
50%         69.900000
75%        129.000000
max       4399.870000
Name: price, dtype: float64

PODEMOS VER QUE TENEMOS 37997 VENTAS DENTRO DE LAS ZONAS QUE GENERAN COMPRAS IGUAL O SUPERIOR AL 1%, DE LOS CUALES TENEMOS
QUE EL VALOR MINIMO DE COMPRA ES DE 1.2 REALES, MIENTRAS QUE EL VALOR MAXIMO ES DE 4399 REALES. 
EL VALOR MEDIO TIENE UN VALOR DE 114.255758 REALES, MIENTRAS QUE LA DESVIACION ESTANDAR TIPICA TIENE UN VALOR DE 172.130391 REALES.
POR ULTIMO PODEMOS VER QUE LO CUARTILES INDICAN QUE EL 25% DE LAS COMPRAS ESTA POR DEBAJO DE UN VALOR DE 38.9 REALES, LUEGO PODEMOS DECIR QUE EL 50% TIENE UN VALOR INFERIOR 69.9 Y EL OTRO 50% UIN VALOR SUPERIOR A 693.9. TAMBIEN INDICA QUE LA MEDIANA ES DE 69.9, YA QUE ES EQUIVALENTE AL 2 CUARTIL.
POR ULTIMO TENEMOS QUE EL 75% DE LAS VENTAS ESTAN POR DEBAJO DE UN VALOR DE 129 REALES Y EL 25% RESTANTE POR ENCIMA DE DICHO VALOR.

In [328]:
cities = df_customer_order_item.groupby("customer_city")["price"].sum()
cities.sort_values(ascending=False)

customer_city
sao paulo         1859466.84
rio de janeiro     955847.36
belo horizonte     346039.04
brasilia           295814.72
curitiba           205999.33
campinas           184657.87
porto alegre       182140.55
salvador           173273.05
guarulhos          138897.20
Name: price, dtype: float64

PODEMOS OBSERVAR QUE EN VISUALIZACION ANTERIOR CUALES SON LOS INGRESOS NONETARIOS EN REALES DE CADA UNA DE LAS LOCALIDADES
QUE GENERAN IGUAL O MAYOR AL 1% DE LAS VENTAS TOTALES. LO QUE SE QUIERE TRABAJAR EN ESTAS LOCALIDADES ES AUMENTAR TANTO 
LA CANTIDAD DE VENTA, COMO ASI TAMBIEN LACANTIDAD DE INGRESOS MONETARIOS. ES DECIR QUE NO SOLO ES AUMENTAR EL NUMERO DE VENTAS,
SI NO TAMBIEN QUE ESAS VENTAS SEAN POR MONTOS MAYORES Y QUE EL CRECIMIENTO MONETARIO CREZCA DE FORMA NOTABLE.
COMPARANDO CON EL ESTUDIO ANTERIOR DE ESTAS 9 REGIONES, DONDE SE CALCULABA CUAL ERA EL NUMERO DE VENTAS, CON ESTE QUE TOMA EL
MONTO DE INGRESO MONETARIO, EL ORDEN EN FORMA DECRECIENTE, VEMOS QUE ES EL MISMO, PERO ESTO PODRIA NO SER ASI, YA QUE PUEDE
PASAR QUE UNA REGION QUE TIENE MENOR NUMERO DE COMPRAS, REALICE COMPRAS POR MONTOS MAYORES, GENERANDO MAYOR INGRESO MONETARIO
QUE OTRA REGION QUE TIENE MAYOR NUMERO DE COMPRAS

##### DF_ORDER_ITEM Y DF_PRODUCTS

In [332]:
df_order_item_product= pd.merge(df_customer_order_item,df_products_mod,how = "inner", on = "product_id")
df_order_item_product.head()

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  128e10d95713541c87cd1a2e48201934  a20e8105f23924cd00833fd87daa0831   
2  0e7e841ddf8f8f2de2bad69267ecfbcf  26c7ac168e1433912a51b924fbd34d34   
3  e6ce16cb79ec1d90b1da9085a6118aeb  494dded5b201313c64ed7f100595b95c   
4  e6ce16cb79ec1d90b1da9085a6118aeb  494dded5b201313c64ed7f100595b95c   

    customer_city  tiempo_entrega_menor_8  order_item_id  \
0       sao paulo                   False              1   
1       sao paulo                    True              1   
2       sao paulo                    True              1   
3  rio de janeiro                   False              1   
4  rio de janeiro                   False              2   

                         product_id  price  product_category_name  
0  87285b34884572647811a353c7ac498a  29.99  utilidades_domesticas  
1  87285b34884572647811a353c7ac498a  29.99  utilidades_domesticas  
2  87285b34884572647811a353c7ac498a  29.99  utilidades_domesticas  
3  08574b074924071f4e201e151b152b4e  99.00     ferramentas_jardim  
4  08574b074924071f4e201e151b152b4e  99.00     ferramentas_jardim

In [334]:
categoria_ingreso = df_order_item_product.groupby("product_category_name")["price"].sum()
categoria_ingreso.sort_values(ascending=False)

product_category_name
beleza_saude                     422810.78
relogios_presentes               381500.32
cama_mesa_banho                  360480.31
esporte_lazer                    323190.57
informatica_acessorios           317337.88
                                   ...    
fashion_esporte                     324.49
artes_e_artesanato                  323.68
fashion_roupa_infanto_juvenil       279.97
pc_gamer                            259.98
seguros_e_servicos                  100.00
Name: price, Length: 74, dtype: float64

PODEMOS VER CUALES SON LAS CATEGORIAS QUE MAYOR INGRESO GENERAN, LOGRANDO CON ESTO PODER GENERAR ESTRATEGIAS DE VENTAS,
YA SEA POR PROMOCIONES O POR DESCUEBTOS O EL METODO QUE MAS SE ACOMODO PARA PODER LLEGAR AL LOGRO FINAL QUE ES PODER GENERAR
MAYOR CANTIDAD DE VENTAS CON UN MAYOR INCREMENTO EN LOS INGRESOS MONETARIOS

In [337]:
categoria_city = df_order_item_product.groupby("customer_city")["product_category_name"].unique()
categoria_city

customer_city
belo horizonte    [ferramentas_jardim, cama_mesa_banho, beleza_s...
brasilia          [cama_mesa_banho, brinquedos, beleza_saude, be...
campinas          [cama_mesa_banho, beleza_saude, livros_interes...
curitiba          [ferramentas_jardim, cama_mesa_banho, construc...
guarulhos         [ferramentas_jardim, informatica_acessorios, e...
porto alegre      [ferramentas_jardim, livros_interesse_geral, e...
rio de janeiro    [ferramentas_jardim, informatica_acessorios, c...
salvador          [ferramentas_jardim, beleza_saude, cool_stuff,...
sao paulo         [utilidades_domesticas, ferramentas_jardim, in...
Name: product_category_name, dtype: object

##### DF_PYMENT Y DF_ORDER_ITEM_PRODUCT

In [340]:
df_order_item_product= pd.merge(df_order_payments_mod,df_order_item_product,how = "inner", on = "order_id")
df_order_item_product

order_id payment_type  payment_installments  \
0      a9810da82917af2d9aefd1278f1dcfa0  credit_card                     1   
1      25e8ea4e93396b6fa0d3dd708e76c1bd  credit_card                     1   
2      771ee386b001f06208a7419e4fc1bbd7  credit_card                     1   
3      595f598849d89203c28d05d0fbf3f92e  credit_card                     1   
4      e647bf617de735448a4ba54f8910393c  credit_card                     1   
...                                 ...          ...                   ...   
39644  20824837882dd673707e0adfcaec1bc1  credit_card                     1   
39645  3531718b24dfd5022fb0d101e10e33b9  credit_card                     1   
39646  7b905861d7c825891d6347454ea7863f  credit_card                     2   
39647  b8b61059626efa996a60be9bb9320e10  credit_card                     5   
39648  28bbae6599b09d39ca406b747b6632b1       boleto                     1   

                            customer_id   customer_city  \
0      f2c7fc58a9de810828715166c672f10a       sao paulo   
1      25b14b69de0b6e184ae6fe2755e478f9       sao paulo   
2      206c0fcc717333d169ff95933fb47341  belo horizonte   
3      cefab390926b6fc3de3ca94c82ebb1ba       sao paulo   
4      11440840f5653ebba13e919285c71389       sao paulo   
...                                 ...             ...   
39644  e59afa77a6682d6efe6297d7d1b1796b  rio de janeiro   
39645  5b819ae762ee1d0424f92fd742a8ed7b  rio de janeiro   
39646  2079230c765a88530822a34a4cec2aa0  belo horizonte   
39647  5d719b0d300663188169c6560e243f27  rio de janeiro   
39648  4c7f868f43b5cff577b0becb8c8b7860  rio de janeiro   

       tiempo_entrega_menor_8  order_item_id  \
0                        True              1   
1                        True              1   
2                       False              1   
3                        True              1   
4                        True              1   
...                       ...            ...   
39644                    True              1   
39645                   False              1   
39646                    True              1   
39647                   False              1   
39648                   False              1   

                             product_id   price        product_category_name  
0      a630cc320a8c872f9de830cf121661a3   17.00                   automotivo  
1      2028bf1b01cafb2d2b1901fca4083222   56.99                   perfumaria  
2      42a2c92a0979a949ca4ea89ec5c7b934   59.90        utilidades_domesticas  
3      c4913f978f6af6755a1dcc8d16e76774   14.85             moveis_decoracao  
4      f03859e4cc21018569f5ae7a03897ff4   49.99               consoles_games  
...                                 ...     ...                          ...  
39644  79ad39409fa4bcc36b4b7f734c79c2a2   69.99  fashion_bolsas_e_acessorios  
39645  db5efde3ad0cc579b130d71c4b2db522  175.00                        audio  
39646  fb55982be901439613a95940feefd9ee   79.00                    papelaria  
39647  e9c1abcc7126360336d8f56d49041a8c  349.00                 beleza_saude  
39648  1cdfaf0cc5abd6063d3a10ae3318f345  169.90              cama_mesa_banho  

[39649 rows x 10 columns]

In [344]:
df_order_item_product["order_id"].duplicated().sum()

6667

In [348]:
df_order_item_product = df_order_item_product.drop_duplicates(subset="order_id")

In [350]:
df_order_item_product.isnull().sum()

order_id                  0
payment_type              0
payment_installments      0
customer_id               0
customer_city             0
tiempo_entrega_menor_8    0
order_item_id             0
product_id                0
price                     0
product_category_name     0
dtype: int64

In [352]:
df_order_item_product = df_order_item_product[["payment_type","payment_installments","customer_city","price","product_category_name"]] 
df_order_item_product.head()

payment_type  payment_installments   customer_city  price  \
0  credit_card                     1       sao paulo  17.00   
1  credit_card                     1       sao paulo  56.99   
2  credit_card                     1  belo horizonte  59.90   
3  credit_card                     1       sao paulo  14.85   
4  credit_card                     1       sao paulo  49.99   

   product_category_name  
0             automotivo  
1             perfumaria  
2  utilidades_domesticas  
3       moveis_decoracao  
4         consoles_games

In [354]:
df_order_item_product["payment_type"].value_counts()

payment_type
credit_card    25607
boleto          5896
voucher          901
debit_card       578
Name: count, dtype: Int64

DE ESTA MANERA PODEMOS VER QUE EXISTEN 4 METODOS DE PAGOS, DE LOS CUALES EL PAGO CON TARJETA DE CREDITO ES EL QUE
MAYORMENTE SE UTILZA, SUPERANDO DE MANERA SIGNIFICATIC AL QUE SE ENCUENTRA ANTERIOR. EXISTEN 25607 COMPRAS

In [358]:
df_order_item_product["payment_installments"].value_counts()

payment_installments
1     16929
2      4317
3      3478
4      2109
10     1520
5      1511
8      1299
6      1114
7       430
9       187
12       38
15       17
11        6
20        6
14        4
18        4
24        4
13        4
21        1
0         1
17        1
16        1
23        1
Name: count, dtype: int64

PODEMOS OBSERVAR QUE LA MAYOR CANTIDAD DE COMPRAS SE REALIZAN EN UNA CUOTA. POR EJEMPLO SE PODRÍA PENSAR EN ESA GENTE
QUE SON QUIENES GENERAN LA MAYOR CANTIDAD DE COMPRAS Y REALIZAR ALGUN TIPO DE PROMOCION DE DESCUENTO, DE ESA MANERA
ES POSIBLE QUE LOS CLIENTES REALICEN COMPRAS DE MAYOR MONTOS, Y QUIENES REALIZAN COMPRAS EN MAYOR CANTIDAD DE PAGOS,
PASEN A REALIZAR COMPRAS EN UN PAGO Y HASTA REALICEN COMPRAS DE MONTOS MAS GRANDES.
PARAS LAS COMPRAS A 6 CUOTAS, ES POSIBLE QUE NO TNGA SENTIDO INVERTIR TIEMPO Y DINERO EN GENERAR PROYECTOS PARA AUMENTAR,
YA QUE NO MODIFICAN EL INGRESO A GRANDES PASOS DE LA EMPRESA

## EXPORTACION DE ARCHIVOS CSV